corehq/apps/sso/models.py

```diff
@@ -118,6 +118,23 @@ def renew_service_provider_certificate(self):
         self.date_sp_rollover_cert_expiration = None
         self.save()
 
+    @property
+    def email_domains(self):
+        return AuthenticatedEmailDomain.objects.filter(
+            identity_provider=self
+        ).values_list('email_domain', flat=True).all()
+
+    @property
+    def sso_exempt_users(self):
+        return UserExemptFromSingleSignOn.objects.filter(
+            email_domain__identity_provider=self,
+        ).values_list('username', flat=True)
```

Comment:
What's the reasoning behind making these methods? The guideline may have changed, but to me, this breaks the principle of least surprise -- properties are lightweight things that are very efficient. If I'm making a database call, that is a function, and shouldn't be disguised as a property. I want to know that there might be some overhead to `get_email_domains()`, for example.

Reply:
the reason is habit. I am referencing these properties in `jinja2` templates, and methods can't be called in templates like `{{ idp.get_email_domains() }}`, though it seems like now `jinja` will call it automatically if it's a function (eg `{{ idp.get_email_domains }}` works), but that wasn't always the case in the past.

Reply:
that being said, I agree with your perspective and I'll update it in a followup PR

Reply:
addressed in https://github.com/dimagi/commcare-hq/pull/29205